src/sentry/debug_files/artifact_bundles.py

```diff
@@ -34,14 +50,374 @@
+            self._merge(artifact_bundle, archive, self.store.identifier.is_indexing_by_release())
+            return True
+        except Exception as e:
+            sentry_sdk.capture_exception(e)
+            metrics.incr("artifact_bundle_flat_file_indexing.error_when_merging")
+            return False
+
+    def _merge(
+        self, artifact_bundle: ArtifactBundle, archive: ArtifactBundleArchive, index_urls: bool
+    ):
+        bundle_index = self._update_bundles(artifact_bundle)
+
+        if index_urls:
+            self._iterate_over_urls(bundle_index, archive)
+        else:
+            self._iterate_over_debug_ids(bundle_index, archive)
+
+    def _update_bundles(self, artifact_bundle: ArtifactBundle) -> int:
+        if artifact_bundle.date_last_modified is None:
+            raise Exception("Can't index a bundle with no date last modified")
+
+        bundle_meta = BundleMeta(
+            id=artifact_bundle.id, timestamp=artifact_bundle.date_last_modified
+        )
+
+        bundle_index = self._index_of_bundle(artifact_bundle.id)
+        if bundle_index is None:
+            self._bundles.append(bundle_meta)
+            return len(self._bundles) - 1
+        else:
+            self._bundles[bundle_index] = bundle_meta
+            return bundle_index
+
+    def _iterate_over_debug_ids(self, bundle_index: int, archive: ArtifactBundleArchive):
+        for debug_id, _ in archive.get_all_debug_ids():
+            self._add_sorted_entry(self._files_by_debug_id, debug_id, bundle_index)
+
+    def _iterate_over_urls(self, bundle_index: int, archive: ArtifactBundleArchive):
+        for url in archive.get_all_urls():
+            self._add_sorted_entry(self._files_by_url, url, bundle_index)
+
+    def _add_sorted_entry(self, collection: Dict[T, List[int]], key: T, bundle_index: int):
+        entries = collection.get(key, [])
+        entries.append(bundle_index)
+        entries.sort(key=lambda index: self._bundles[index]["timestamp"], reverse=True)
+        collection[key] = entries
+
+    def remove(self, artifact_bundle_id: int) -> bool:
+        try:
+            self._remove(artifact_bundle_id)
+            return True
+        except Exception as e:
+            sentry_sdk.capture_exception(e)
+            metrics.incr("artifact_bundle_flat_file_indexing.error_when_removing_bundle")
+            return False
+
+    def _remove(self, artifact_bundle_id: int):
+        bundle_index = self._index_of_bundle(artifact_bundle_id)
+        if bundle_index is not None:
+            self._files_by_url = self._update_bundle_references(self._files_by_url, bundle_index)
+            self._files_by_debug_id = self._update_bundle_references(
+                self._files_by_debug_id, bundle_index
+            )
+            self._bundles.pop(bundle_index)
+        else:
+            raise Exception("The bundle you want do delete doesn't exist in the index.")
```

Comment:
IMO error handling should move to the consumer.

The bundle index internally does not care if you are trying to remove something that was never part of the index.
Also, you are already returning `True/`False` to the consumer, so that code can deal with metrics or error handling as it sees fit.

Same thing for `build` (aka load) and `merge` as well.

Reply:
I was thinking about this also and I agree, will lift out error handling.